xbmc/platform/darwin/osx/XBMCApplication.mm

```diff
@@ -0,0 +1,390 @@
+                           withObject:nil];
+
+  // Set the working directory to the .app's parent directory
+  [self setupWorkingDirectory];
+
+  [NSWorkspace.sharedWorkspace.notificationCenter addObserver:self
+                                                     selector:@selector(deviceDidMountNotification:)
+                                                         name:NSWorkspaceDidMountNotification
+                                                       object:nil];
+
+  [NSWorkspace.sharedWorkspace.notificationCenter
+      addObserver:self
+         selector:@selector(deviceDidUnMountNotification:)
+             name:NSWorkspaceDidUnmountNotification
+           object:nil];
+
+  // Hand off to main application code
+  gCalledAppMainline = TRUE;
+
+  //window.acceptsMouseMovedEvents = TRUE;
+
+  // kick our mainloop into an extra thread
+  [NSThread detachNewThreadSelector:@selector(mainLoopThread:) toTarget:self withObject:nil];
+}
+
+- (NSApplicationTerminateReply)applicationShouldTerminate:(NSApplication*)sender
+{
+  return NSTerminateNow;
+}
+
+- (void)applicationWillTerminate:(NSNotification*)note
+{
+  [NSWorkspace.sharedWorkspace.notificationCenter removeObserver:self
+                                                            name:NSWorkspaceDidMountNotification
+                                                          object:nil];
+
+  [NSWorkspace.sharedWorkspace.notificationCenter removeObserver:self
+                                                            name:NSWorkspaceDidUnmountNotification
+                                                          object:nil];
+}
+
+- (void)applicationWillResignActive:(NSNotification*)note
+{
+  // when app moves to background
+}
+
+- (void)applicationWillBecomeActive:(NSNotification*)note
+{
+  // when app moves to front
+}
+
+/*
+ * Catch document open requests...this lets us notice files when the app
+ *  was launched by double-clicking a document, or when a document was
+ *  dragged/dropped on the app's icon. You need to have a
+ *  CFBundleDocumentsType section in your Info.plist to get this message,
+ *  apparently.
+ *
+ * Files are added to gArgv, so to the app, they'll look like command line
+ *  arguments. Previously, apps launched from the finder had nothing but
+ *  an argv[0].
+ *
+ * This message may be received multiple times to open several docs on launch.
+ *
+ * This message is ignored once the app's mainline has been called.
+ */
+
+//! @Todo Transition to application: openURLs:
+- (BOOL)application:(NSApplication*)theApplication openFile:(NSString*)filename
+{
+  const char* temparg;
+  size_t arglen;
+  char* arg;
+  char** newargv;
+
+  // app has started, ignore this document.
+  if (gCalledAppMainline)
+    return FALSE;
+
+  temparg = [filename UTF8String];
+  arglen = strlen(temparg) + 1;
+  arg = (char*)malloc(arglen);
+  if (arg == NULL)
+    return FALSE;
+
+  newargv = (char**)realloc(gArgv, sizeof(char*) * (gArgc + 2));
```

Comment:
This whole method looks like it could use some love but this looks like a great use for std::vector instead of malloc/realloc

Reply:
Its already flagged for a todo to replace with a different method. Will leave it as is for now, as it'll be removed completely and replaced with an alternative function (at some point).